File: Loenn/entities/customFlutterbird.lua
local jautils = require("mods").requireFromPlugin("libraries.jautils")
local drawableSpriteStruct = require("structs.drawable_sprite")
local utils = require("utils")

local texture = "scenery/flutterbird/idle00"
local flutterbird = {}

flutterbird.name = "FrostHelper/CustomFlutterBird"
flutterbird.depth = -9999

jautils.createPlacementsPreserveOrder(flutterbird, "normal", {
    { "colors", "89fbff,f0fc6c,f493ff,93baff", "colorList" },
    { "directory", "scenery/flutterbird/" },
    { "hopSfx", "event:/game/general/birdbaby_hop" },
    { "flyAwaySfx", "event:/game/general/birdbaby_flyaway" },
    { "tweetingSfx", "event:/game/general/birdbaby_tweet_loop" },
    { "dontFlyAway", false },
})

function flutterbird.sprite(room, entity)
    utils.setSimpleCoordinateSeed(entity.x, entity.y)
    local colors = jautils.getColors(entity.colors)
    local colorIndex = math.random(1, #colors)

    local flutterbirdSprite = drawableSpriteStruct.fromTexture(texture, entity)

    flutterbirdSprite:setJustification(0.5, 1.0)
    flutterbirdSprite:setColor(colors[colorIndex])

    return flutterbirdSprite
end

return flutterbird

File: Ahorn/entities/customFlutterbird.jl
module FrostHelperFlutterBird

using ..Ahorn, Maple

@mapdef Entity "FrostHelper/CustomFlutterBird" FHCustomFlutterbird(x::Integer, y::Integer, colors::String="89fbff,f0fc6c,f493ff,93baff", dontFlyAway::Bool=false,
                                                                   hopSfx::String="event:/game/general/birdbaby_hop", flyAwaySfx::String="event:/game/general/birdbaby_flyaway",
                                                                   tweetingSfx::String="event:/game/general/birdbaby_tweet_loop")

const placements = Ahorn.PlacementDict(
    "Custom Flutterbird (Frost Helper)" => Ahorn.EntityPlacement(
        FHCustomFlutterbird
    ),
)

const sprite = "scenery/flutterbird/idle00.png"

const colors = [
    (137, 251, 255, 255) ./ 255,
    (240, 252, 108, 255) ./ 255,
    (244, 147, 255, 255) ./ 255,
    (147, 186, 255, 255) ./ 255
]

function Ahorn.selection(entity::FHCustomFlutterbird)
    x, y = Ahorn.position(entity)

    return Ahorn.getSpriteRectangle(sprite, x, y, jx=0.5, jy=1.0)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::FHCustomFlutterbird, room::Maple.Room)
    rng = Ahorn.getSimpleEntityRng(entity)
    color = rand(rng, colors)

    Ahorn.drawSprite(ctx, sprite, 0, 0, jx=0.5, jy=1.0, tint=color)
end

end